docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,450 @@
+     injection point.
+   - The controller will now identify all in-cluster objects in the same
+     namespace as the PackageVariant resource, with GVK matching the injection
+     point (the in-package resource).
+   - The controller will look through the list of injection selectors in
+     order and checking if any of the in-cluster objects match the selector. If
+     so, that in-cluster object is selected, and processing of the list of
+     injection selectors stops. Note that in the initial version, the selectors
+     are only by name; thus, at most one match is possible for any given
+     selector.
+   - If no in-cluster object is selected, the injection point ConditionType will
+     be set to `False` with a message that no matching in-cluster resource was
+     found, and processing proceeds to the next injection point.
+   - If a matching in-cluster object is selected, then it is injected as
+     follows:
+     - The `spec` field from the in-cluster resource is copied to the `spec`
+       field of the in-package resource (the injection point), overwriting it.
+     - An annotation with name `kpt.dev/injected-resource-name` and value set to
+       the name of the in-cluster resource is added (or overwritten) in the
+       in-package resource.
+1. The PackageVariant resource must have a condition of type `ConfigInjected`.
+   This will be set to `True` if and only if:
+   - All configuration injection processing is complete.
+   - There is no resource annotated as an injection point but having an invalid
+     annotation value (i.e., other than `required` or `optional`).
+   - Matching in-cluster resources were successfully injected for all `required`
+     injection points (all required injection point ConditionTypes are `True`).
+   - There are no ambiguous condition types due to conflicting GVK and name
+     values. These must be disambiguated in the upstream package, if so.
+
+Note that by allowing the use of GVK, not just name, in the selector, more
+precision in selection is enabled. This is a way to constrain the injections
+that will be done. That is, if the package has 10 different objects with
+`config-injection` annotation, the PackageVariant could say it only wants to
+replace certain GVKs, allowing better control.
+
+Consider, for example, if the cluster contains these resources:
+
+- GVK1 foo
+- GVK1 bar
+- GVK2 foo
+- GVK2 bar
+
+If we could only define injection selectors based upon name, it would be
+impossible to ever inject one GVK with `foo` and another with `bar`. Instead,
+by using GVK, we can accomplish this with a list of selectors like:
+
+ - GVK1 foo
+ - GVK2 bar
+
+That said, often name will be sufficiently unique when combined with the
+in-package resource GVK, and so making the selector GVK optional is more
+convenient.
+
+#### Order of Mutation During Creation
+Note that the KRM function pipeline defined in the Kptfile runs every time Porch
```

Comment:
@natasha41575 can you take a look at this? Obviously we haven't implemented this yet, but I am wondering if this is the expected behavior. Would we save the package after each step, or just at the end of the reconciliation? If it's the latter, then this becomes:

1. Package Context Injections
1. PackageVariant KRM Function Pipeline
1. Config Injection
1. Kptfile KRM Function Pipeline

WDYT?

Reply:
Yeah, I think that's correct. I'll update the design.

Reply:
done